File: frontend/src/app/services/evaluation-activite.service.ts
// src/app/services/evaluation-activite.service.ts — MISE À JOUR
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import {
  EvaluationActivite,
  EvaluationActiviteRequest,
  EvaluationsActiviteResponse,
  EvaluationResume
} from '../shared/models/evaluationactivite.model';

@Injectable({ providedIn: 'root' })
export class EvaluationActiviteService {

  private http = inject(HttpClient);
  private base = 'http://localhost:8085/api/application/activites';

  getByActivite(activiteId: number): Observable<EvaluationsActiviteResponse> {
    return this.http.get<EvaluationsActiviteResponse>(
        `${this.base}/${activiteId}/evaluations`);
  }

  /**
   * ✅ NOUVEAU — Résumé (moyenne + total) pour TOUTES les activités d'un
   * projet en un seul appel, utilisé par la colonne "Évaluation" de la
   * table/kanban (évite un appel par activité).
   */
  getResumeByProjet(projetId: number): Observable<EvaluationResume[]> {
    return this.http.get<EvaluationResume[]>(
        `${this.base}/projet/${projetId}/evaluations-resume`);
  }

  /** Upsert : crée l'évaluation de l'utilisateur courant si absente, sinon la met à jour. */
  evaluer(activiteId: number, req: EvaluationActiviteRequest): Observable<EvaluationActivite> {
    return this.http.post<EvaluationActivite>(
        `${this.base}/${activiteId}/evaluations`, req);
  }

  /** Supprime UNIQUEMENT l'évaluation de l'utilisateur courant (vérifié côté backend). */
  supprimer(activiteId: number): Observable<void> {
    return this.http.delete<void>(`${this.base}/${activiteId}/evaluations`);
  }
}

File: backend/application-service/src/main/java/com/talenthub/application_service/Service/EvaluationActiviteService.java
// Service/EvaluationActiviteService.java — MISE À JOUR
package com.talenthub.application_service.Service;

import com.talenthub.application_service.DTO.EvaluationResumeDTO;
import com.talenthub.application_service.Entity.Activité;
import com.talenthub.application_service.Entity.EvaluationActivite;
import com.talenthub.application_service.Exception.ResourceNotFoundException;
import com.talenthub.application_service.Repository.ActiviteRepository;
import com.talenthub.application_service.Repository.EvaluationActiviteRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Map;

@Slf4j
@Service
@RequiredArgsConstructor
public class EvaluationActiviteService {

    private final EvaluationActiviteRepository evaluationRepo;
    private final ActiviteRepository           activiteRepo;

    @Transactional(readOnly = true)
    public List<EvaluationActivite> getByActivite(Long activiteId) {
        return evaluationRepo.findByActiviteId(activiteId);
    }

    @Transactional(readOnly = true)
    public Double getMoyenne(Long activiteId) {
        Double avg = evaluationRepo.getMoyenneNoteByActiviteId(activiteId);
        return avg != null ? Math.round(avg * 10) / 10.0 : null;
    }

    @Transactional(readOnly = true)
    public long countByActivite(Long activiteId) {
        return evaluationRepo.countByActiviteId(activiteId);
    }

    /**
     * ✅ NOUVEAU — Résumé (moyenne + total) de TOUTES les activités d'un
     * projet en une seule requête, pour la colonne "Évaluation" de la
     * table/kanban côté frontend.
     */
    @Transactional(readOnly = true)
    public List<EvaluationResumeDTO> getResumeByProjet(Long projetId) {
        return evaluationRepo.getResumeByProjetId(projetId).stream()
                .map(row -> new EvaluationResumeDTO(
                        (Long) row[0],
                        (Double) row[1],
                        ((Number) row[2]).longValue()))
                .toList();
    }

    /**
     * Upsert : si l'évaluateur a déjà noté cette activité, on met à jour sa
     * note/commentaire existant (contrainte unique activite+évaluateur).
     * Sinon, on crée une nouvelle évaluation.
     */
    @Transactional
    public EvaluationActivite evaluer(Long activiteId, String evaluateurKeycloakId,
                                      String evaluateurNom, Integer note, String commentaire) {
        if (note == null || note < 0 || note > 5) {
            throw new IllegalArgumentException("La note doit être comprise entre 0 et 5.");
        }

        Activité activite = activiteRepo.findById(activiteId)
                .orElseThrow(() -> new ResourceNotFoundException("Activité non trouvée: " + activiteId));

        EvaluationActivite eval = evaluationRepo
                .findByActiviteIdAndEvaluateurKeycloakId(activiteId, evaluateurKeycloakId)
                .orElseGet(() -> EvaluationActivite.builder()
                        .activite(activite)
                        .evaluateurKeycloakId(evaluateurKeycloakId)
                        .build());

        eval.setEvaluateurNom(evaluateurNom);
        eval.setNote(note);
        eval.setCommentaire(commentaire);

        EvaluationActivite saved = evaluationRepo.save(eval);
        log.info("⭐ Évaluation activité id={} par {} → note={}", activiteId, evaluateurKeycloakId, note);
        return saved;
    }

    /**
     * Supprime l'évaluation d'un évaluateur précis pour une activité précise.
     * Un évaluateur ne peut supprimer que SA PROPRE évaluation (vérifié au
     * niveau du contrôleur via le keycloakId du token, pas ici).
     */
    @Transactional
    public void supprimer(Long activiteId, String evaluateurKeycloakId) {
        evaluationRepo.deleteByActiviteIdAndEvaluateurKeycloakId(activiteId, evaluateurKeycloakId);
    }

    public static Map<String, Object> toResume(Double moyenne, long total) {
        return Map.of(
                "moyenne", moyenne != null ? moyenne : 0.0,
                "total", total
        );
    }
}